charmstore.go

```diff
@@ -232,6 +233,94 @@ func (s *CharmStore) Latest(curls ...*charm.URL) ([]CharmRevision, error) {
 	return responses, nil
 }
 
+func (s *CharmStore) ListResources(ids ...*charm.URL) ([]ResourceResult, error) {
+	if len(ids) == 0 {
+		return nil, nil
+	}
+
+	results, err := s.client.ListResources(ids...)
+	if err != nil {
+		return nil, errgo.NoteMask(err, "cannot get resource metadata from the charm store", errgo.Any)
+	}
+
+	result := make([]ResourceResult, len(ids))
+	for i, id := range ids {
+		resources, ok := results[id.String()]
+		if !ok {
+			result[i].Err = CharmNotFound(id.String())
+			continue
+		}
+		list := make([]resource.Resource, len(resources))
+		for j, res := range resources {
+			resType, err := apiResourceType2ResourceType(res.Type)
+			if err != nil {
+				return nil, errgo.Notef(err, "bad data returned for resource %q", res.Name)
+			}
+			fp, err := resource.ParseFingerprint(res.Fingerprint)
+			if err != nil {
+				return nil, errgo.Notef(err, "bad data returned for resource %q", res.Name)
+			}
+			list[j] = resource.Resource{
+				Meta: resource.Meta{
+					Name:        res.Name,
+					Type:        resType,
+					Path:        res.Path,
+					Description: res.Description,
+				},
+				Revision:    res.Revision,
+				Fingerprint: fp,
+				Size:        res.Size,
+			}
+		}
+		result[i].Resources = list
+	}
+	return result, nil
+}
+
+func (c *CharmStore) UploadResource(id *charm.URL, name, filename string) (revision int, err error) {
```

Comment:
This is just the metadata.. there's no io.Reader, unless I'm not understanding something.